File: scala/src/main/scala/org/agilegrenoble/objectcalisthenics/GildedRose.scala
package org.agilegrenoble.objectcalisthenics

object GildedRose {

  def main(args: Array[String]) {
    new GildedRose().updateQuality()
  }
  
  
  
}

class GildedRose {

  var items =
        List(Item("+5 Dexterity Vest", 10, 20),
        Item("Aged Brie", 2, 0),
        Item("Elixir of the Mongoose", 5, 7),
        Item("Sulfuras, Hand of Ragnaros", 0, 80),
        Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
        Item("Conjured Mana Cake", 3, 6))

  def updateQuality() {
    for (i <- 0 until items.size) {
      if (items(i).name != "Aged Brie" && 
        items(i).name != "Backstage passes to a TAFKAL80ETC concert") {
        if (items(i).quality > 0) {
          if (items(i).name != "Sulfuras, Hand of Ragnaros") {
            items(i).quality= items(i).quality -1
          }
        }
      } else {
        if (items(i).quality < 50) {
          items(i).quality=items(i).quality+1
          if (items(i).name == "Backstage passes to a TAFKAL80ETC concert") {
            if (items(i).sellIn < 11) {
              if (items(i).quality < 50) {
                items(i).quality=(items(i).quality + 1)
              }
            }
            if (items(i).sellIn < 6) {
              if (items(i).quality < 50) {
                items(i).quality=(items(i).quality + 1)
              }
            }
          }
        }
      }
      if (items(i).name != "Sulfuras, Hand of Ragnaros") {
        items(i).sellIn=(items(i).sellIn - 1)
      }
      if (items(i).sellIn < 0) {
        if (items(i).name != "Aged Brie") {
          if (items(i).name != "Backstage passes to a TAFKAL80ETC concert") {
            if (items(i).quality > 0) {
              if (items(i).name != "Sulfuras, Hand of Ragnaros") {
                items(i).quality=(items(i).quality - 1)
              }
            }
          } else {
            items(i).quality=(items(i).quality - items(i).quality)
          }
        } else {
          if (items(i).quality < 50) {
            items(i).quality=(items(i).quality + 1)
          }
        }
      }
    }
  }
}

File: c#/GuildedRose/GildedRose.cs
using System;
using System.Collections.Generic;

public class GildedRose
{

    //public static void Main(String[] args) {
    //    GildedRose gildedRose = new GildedRose();
    //    gildedRose.UpdateQuality(gildedRose.makeItems());
    //}

    public List<Item> makeItems()
    {
        List<Item> items = new List<Item>();
        items.Add(new Item("+5 Dexterity Vest", 10, 20));
        items.Add(new Item("Aged Brie", 2, 0));
        items.Add(new Item("Elixir of the Mongoose", 5, 7));
        items.Add(new Item("Sulfuras, Hand of Ragnaros", 0, 80));
        items.Add(new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20));
        items.Add(new Item("Conjured Mana Cake", 3, 6));
        return items;
    }

    public void UpdateQuality(List<Item> list)
    {
        List<Item> items = list;
        foreach (var item in items)
        {
            if (item.Name != "Aged Brie" && item.Name != "Backstage passes to a TAFKAL80ETC concert")
            {
                if (item.Quality > 0)
                {
                    if (item.Name != "Sulfuras, Hand of Ragnaros")
                    {
                        item.Quality = item.Quality - 1;
                    }
                }
            }
            else
            {
                if (item.Quality < 50)
                {
                    item.Quality = item.Quality + 1;

                    if (item.Name.Equals("Backstage passes to a TAFKAL80ETC concert"))
                    {
                        if (item.SellIn < 11)
                        {
                            if (item.Quality < 50)
                            {
                                item.Quality = item.Quality + 1;
                            }
                        }

                        if (item.SellIn < 6)
                        {
                            if (item.Quality < 50)
                            {
                                item.Quality = item.Quality + 1;
                            }
                        }
                    }
                }
            }

            if (!item.Name.Equals("Sulfuras, Hand of Ragnaros"))
            {
                item.SellIn = item.SellIn - 1;
            }

            if (item.SellIn < 0)
            {
                if (item.Name != "Aged Brie")
                {
                    if (item.Name != "Backstage passes to a TAFKAL80ETC concert")
                    {
                        if (item.Quality > 0)
                        {
                            if (item.Name != "Sulfuras, Hand of Ragnaros")
                            {
                                item.Quality = item.Quality - 1;
                            }
                        }
                    }
                    else
                    {
                        item.Quality = item.Quality - item.Quality; 
                    }
                }
                else
                {
                    if (item.Quality < 50)
                    {
                        item.Quality = item.Quality + 1;
                    }
                }
            }
        }
    }

}
